### PR TITLE
WIP: Update maintainers guide

### PR DESCRIPTION
**Pull request by tux-rampage**
|    Q          |   A
|-------------- | ------
| Documentation | yes
| Bugfix        | no
| BC Break      | yes?
| New Feature   | no
| RFC           | yes
| QA            | yes

### Description

As discussed with @weierophinney and the contributors channel, the maintainers guide needs to be updated to the automatic releases workflow.

This PR is just a starting point to drive this document with the technical steering committee to a guide that helps maintainers to maintain the component as intended and have a point of orientation to refer to.

I just incorporated the new process as much as I understood it - which might be off or even totally wrong. So this might also help me and potential other maintainers, who are not part of the steering committee, to better understand the intended process.

### CODE REVIEW

File: MAINTAINERS.md [hunk 1]
```diff
@@ -23,37 +23,22 @@ matching the repository name:
...
 
-- a **master** branch. This branch MUST be releasable at all times. Commits and merges against
-  this branch MUST contain only bugfixes and/or security fixes. Maintenance releases are tagged
-  against master.
-- a **develop** branch. This branch contains *new features*, and will either become the next minor
-  (feature) release or next major release. Typically, major releases are reserved for backwards
-  *incompatible* changes, but can also be used to signal major new features.
+- two or more minor version branches of the current major release
+  - the highest minor version branch must not be released (i.e. does not have matching version tags like `1.0.0` 
+    for `1.0.x`).
```
Comment (weierophinney) [inline MAINTAINERS.md:31]: This part isn't entirely true or necessary. There can be a single branch with no releases existing (e.g., a 0.1.x branch), or where the migration encountered a `develop` branch that had no changes from the `master` branch, and we only published the release branch associated with the most recent release.

AFTER a release has been made using the automatic-releases workflow, there will always be a minimum of two release branches. But BEFORE, there may only be one.

File: MAINTAINERS.md [hunk 2]
```diff
@@ -23,37 +23,22 @@ matching the repository name:
...
+  - all other minor version branches must be released (i.e. have one or more version tags).
+- only unrelased branches may receive *new features*.
```
Comment (weierophinney) [inline MAINTAINERS.md:33]:
```suggestion
- only unreleased branches may receive *new features*.
```

File: MAINTAINERS.md [hunk 3]
```diff
@@ -23,37 +23,22 @@ matching the repository name:
...
->
-> Because we follow git flow, any changes made on `master` must also be merged to `develop`, and
-> this cannot be accomplished from the web interface. Additionally, we request that you annotate
-> the merge commits to reference the pull request and any associated issues; this practice
-> simplifies determining the source of a feature, and allows developers to discover the rationale
-> as well as any discussion that occurred.
->
-> The remainder of this document details how to merge pull requests to the repositories.
+released immediately and coordinated with the TSC to ensure an advisory is made, a
+CVE is created, and that the fix is backported to the Long Term Support releases.
```
Comment (weierophinney) [inline MAINTAINERS.md:41]: This part changes.

We don't have LTS anymore.

We have "security windows". Essentially, in the case of a security advisory and associated release(s), we make the patch to the earliest release branch we still support. From there, automatic-releases takes over, and does merge-up releases to each next latest release branch until it hits the latest. Each of those releases along the way will be listed in the security advisory.

File: MAINTAINERS.md [hunk 4]
```diff
@@ -135,16 +120,15 @@ During your review, consider the following points:
...
 
-Check which branch the PR was made against — patches made against the `develop` branch should only
-be merged to `develop`. Patches made against the `master` branch may be merged to only `develop`, or
-both `develop` AND `master`. The criteria to use in this latter situation is: does the patch
-introduce anything new to the API? does it potentially mark a change in behavior of any sort? If the
-answer to either of these is "yes", merge only to the `develop` branch.
+Pull Requests should only be merged into the branches they were made against. When a PR contains a 
+feature and was made against an already released branch, you should work with the contributor to target
+it against the next release branch. The PR should have a Milestone matching the next release of the
+target branch (for example milestone `1.0.1` for branch `1.0.x`).
```
Comment (weierophinney) [inline MAINTAINERS.md:126]: This workflow works pretty much the opposite now.

Since the default branch is the latest release branch, the more common scenario is that somebody will submit a bugfix against the latest release branch, when it should go against the previous. In those cases, you'll need to re-target the PR and help the user rebase or cherry-pick against the previous release branch.

File: MAINTAINERS.md [hunk 5]
```diff
@@ -135,16 +120,15 @@ During your review, consider the following points:
...
 
 To sum up:
 
-- If the commit is a bugfix, merge to **both** `master` **and** `develop`.
-- If the commit is a new feature, merge to `develop` **only**.
+- If the PR is a bugfix, it should target the lowest affected release branch
```
Comment (weierophinney) [inline MAINTAINERS.md:130]: I think we only want to target the lowest affected release branch for _security fixes_. For bugfixes, the latest release branch with a release created from it would be enough.

So, as an example, let's say we had a repo with the following release branches:

- 2.7.x
- 2.8.x
- 2.9.x

and tags:

- 2.7.0
- 2.7.1
- 2.7.2
- 2.8.0
- 2.8.1

We'd apply a bugfix against the 2.8.x branch in that case, and then end up issuing a 2.8.2 release.

File: MAINTAINERS.md [hunk 6]
```diff
@@ -135,16 +120,15 @@ During your review, consider the following points:
...
+- If the PR is a new feature, merge to `develop` **only**.
```
Comment (weierophinney) [inline MAINTAINERS.md:131]:
```suggestion
- If the PR is a new feature, merge to the latest existing release branch against which no tags have been created.
  If none exists, create a new one from the latest existing release branch using `git switch -c {X.Y+1.x} {X.Y.x}`.
```

**Review by weierophinney — CHANGES_REQUESTED** [review]
Awesome start, thank you!

I've left a few notes, as well as a few code suggestions.

File: MAINTAINERS.md [hunk 7]
```diff
@@ -23,37 +23,29 @@ matching the repository name:
 This guide is intended for *maintainers* — anybody with commit access to one or more laminas Framework
 repositories.
 
-We use the [git flow methodology](http://nvie.com/posts/a-successful-git-branching-model/) for
+We use [automatic relases](https://github.com/laminas/automatic-releases) for
```
Comment (weierophinney) [inline MAINTAINERS.md:26]:
```suggestion
We use [automatic releases](https://github.com/laminas/automatic-releases) for
```

File: MAINTAINERS.md [hunk 8]
```diff
@@ -23,37 +23,29 @@ matching the repository name:
...
 managing the various laminas Framework repositories. At a glance, this means:
```
Comment (weierophinney) [inline MAINTAINERS.md:27]: Should be "Laminas Project", not "laminas Framework".

File: MAINTAINERS.md [hunk 9]
```diff
@@ -132,67 +130,27 @@ During your review, consider the following points:
...
+  The only exception is when the package is not yet released and therefore has only 
+  one release branch.
+- If the PR is a new feature, merge it to the latest release branch **only**.
 
-Now you're staged and ready to pull the changeset in. On the bar that occurs before comments, on the
-left side is a little "information" icon. Click on this, and you get a popup dialog showing the
-steps to take to manually merge. If you click the icon next to the second step, it will copy a "git
-pull" command to your clipboard. Paste this into your terminal.
+You may merge it by using the green "Merge" button in github. If you do so, use only the
+"Create a merge commit" option. Do **not** use "Squash and merge" or "Rebase and merge".
```
Comment (weierophinney) [inline MAINTAINERS.md:143]: I don't think we we ever discussed this in the TSC. What's the rationale for "create a merge commit" as the selected option? "Rebase and merge" leads to a more linear tree, which makes granular reversions of changes easier, while "Squash and merge" makes it easy to back out the entire changeset at once (though I'd argue squash and merge is _never_ a good idea because you lose granularlity). I've often used the `merge --ff` option when manually merging, which is similar to "Rebase and merge".

Maybe @Ocramius could weigh in?

File: MAINTAINERS.md [hunk 10]
```diff
@@ -132,67 +130,27 @@ During your review, consider the following points:
...
-> GitHub develops a [hub](https://github.com/github/hub) command that provides a superset of
-> features for your git executable. It can make the following far simpler:
->
-> ```console
-> $ git merge <url to PR>
-> ```
->
-> We recommend using the hub command to make your life as a maintainer easier!
+Alternatively you may clone the repository and merge the PR locally. If you decide to do so, 
+do not apply additional changes other than changelog entries to the merge.
```
Comment (weierophinney) [inline MAINTAINERS.md:146]: This doesn't detail _how_ to merge. Should it be a standard merge, with a merge commit? Or a `merge --ff`? Or should we rebase and then `merge --ff`? (This discussion is related to my comments about using the button for merging. We need consistency, but also a good rationale for what we choose. And it needs to be the same whether using the button, or doing it at the CLI.)

File: MAINTAINERS.md [hunk 11]
```diff
@@ -222,104 +180,15 @@ we follow [Keep a CHANGELOG](http://keepachangelog.com/). The format is simple:
...
-> ```dosini
-> [branch "master"]
->     ...
->     mergeoptions = --no-ff
-> ```
-
-As an example of merging a hotfix:
+noted; things like bugfixes, coding standards fixes, continuous integration changes, or typo fixes do not need
+to be communicated. However, anything that falls under an addition, deprecation, removal
+MUST be noted. Please provide a succinct but *narrative* description for the change you merge.
```
Comment (weierophinney) [inline MAINTAINERS.md:185]:
```suggestion
noted; things like coding standards fixes, continuous integration changes, or typo fixes do not need
to be communicated. However, any addition, deprecation, removal, or fix with user impact
SHOULD be noted. Please provide a succinct but *narrative* description for the change you merge,
detailing the change, its impact on the user, and what changes they may need to make to their code.
```

File: MAINTAINERS.md [hunk 12]
```diff
@@ -222,104 +180,15 @@ we follow [Keep a CHANGELOG](http://keepachangelog.com/). The format is simple:
...
-> ```console
-> $ git checkout develop && git reset --hard origin/develop
-> $ git checkout master && git reset --hard origin/master
-> $ git merge --no-ff hotfix/2719
-> $ git checkout develop
-> $ git merge --no-ff hotfix/2719
-> ```
->
-> If you notice merge conflicts, delete your merge branch, and start again from the top.
+Releases are automatically performed by closing a milestone.
```
Comment (weierophinney) [inline MAINTAINERS.md:191]:
```suggestion
Releases are automatically performed by closing a milestone. All issues and pull
requests in the milestone MUST be closed before closing the milestone; otherwise,
the release will not be issued.
```

**Review by weierophinney — CHANGES_REQUESTED** [review]
A few more changes, and a couple points for discussion.

File: MAINTAINERS.md (hunk 9 above)
Comment (tux-rampage) [reply MAINTAINERS.md:143]: Actually this is just a suggestion since there is official statement here and I tried to base this on the previous gitflow strategy. But that all is a matter of taste and whatever is deemed to be important about the commit history. 

- A merge commit makes the origin of a change more obvious in the commit graph IMHO. 
- A squashed commit will consider the PR as a single commit which is sometimes useful to get rid of small adjustment commits during review rounds. It requires discipline to create small changesets though (which is always encouraged from a maintainers pov) - but this might drive off some contributors that are not familiar with this process.
- With a rebase and merge you'd loose the ability to track commits for its origin, but it produces a linear commit history.

I could also rephrase it to a "Pick whatever suits the individual situation for the PR" - which might be the best option here.

File: MAINTAINERS.md (hunk 10 above)
Comment (tux-rampage) [reply MAINTAINERS.md:146]: That's right much like above I actually don't know how to start here. At first do we want a permissive approach or restrictive approach. Both has its pros and cons. Permissive offers more flexibility, restrictive is more streamlined. Whatever the descition is: Here it should be clear to a Maintainer, what's good to go and what should not be done in either a permissive ("Don't do x, y, z - otherwise do as you think it suits best") or restrictive ("Only do x, y, z") way.

**Comment by tux-rampage** [comment]
@weierophinney thanks for taking time to review - I'll pull in your changes, but not sure when I'll be able to do so. I'm quite packed with contractor work atm.

**Comment by weierophinney** [comment]
@tux-rampage I've pushed changes based on feedback I provided, some feedback I obtained from this week's TSC meeting, and a complete review of the document (I was able to greatly simplify the release and changelog sections). I've also moved to one-sentence-per-line structure, which is what we're aiming for in the documentation going forwards.

I'm going to open this for review, and ask other TSC members to take a look.

**Review by Ocramius — APPROVED** [review]
Skimmed, and I think overall it makes sense. Especially like the new nomination template idea :+1:

**Review by weierophinney — APPROVED** [review]
